src/main/java/htsjdk/samtools/util/BlockCompressedIndex.java

```diff
@@ -0,0 +1,248 @@
+     */
+    public void writeIndex(final Path output) throws IOException {
+        if (output == null) {
+            throw new IllegalArgumentException("null output path");
+        }
+
+        final ByteBuffer buffer = allocateBuffer(entries.length, true);
+
+        // put the number of entries
+        buffer.putLong(Integer.toUnsignedLong(entries.length));
+
+        for (final IndexEntry entry : entries) {
+            // TODO - check if it is unsigned?
+            buffer.putLong(entry.getCompressedOffset());
+            buffer.putLong(entry.getUncompressedOffset());
+        }
+
+        // write into the output
+        Files.write(output, buffer.array());
+    }
+
+    /**
+     * Load the index from the provided file.
+     *
+     * @param indexPath the path for the index to load.
+     *
+     * @return loaded index.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public static final BlockCompressedIndex loadIndex(final Path indexPath) throws IOException {
+        // allocate a buffer for re-use for read each byte
+        ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
+        buffer.order(ByteOrder.LITTLE_ENDIAN);
+
+        try (final ByteChannel channel = Files.newByteChannel(indexPath)) {
+            if (Long.BYTES != channel.read(buffer)) {
+                throw new IOException("corrupted index file");
+            }
+            buffer.flip();
+            final int numberOfEntries = (int) buffer.getLong();
+
+            // allocate array with the entries
+            final IndexEntry[] entries = new IndexEntry[numberOfEntries];
+
+            // create a new buffer with the correct size and read into it
+            buffer = allocateBuffer(numberOfEntries, false);
+            channel.read(buffer);
+            buffer.flip();
+
+            for (int i = 0; i < numberOfEntries; i++) {
+                entries[i] = new IndexEntry(buffer.getLong(), (int) buffer.getLong());
+            }
+
+            return new BlockCompressedIndex(entries);
+        }
+    }
+
+    /**
+     * Creates an index for a block-compressed file.
+     *
+     * @param bgzipFile the bqzip file.
+     *
+     * @return generated index.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public static final BlockCompressedIndex createIndex(final Path bgzipFile)
+            throws IOException {
+        if (bgzipFile == null) {
+            throw new IllegalArgumentException("null input path");
+        }
+        // open the file for reading as a block-compressed file
+        final BlockCompressedInputStream bgzipStream =
+                new BlockCompressedInputStream(Files.newInputStream(bgzipFile));
+
+        // store the entries as a list
+        final List<IndexEntry> entries = new ArrayList<>();
+
+        // we need to track how many bytes we read to compute the offset
+        // because the at the end of the block, the offset is set to 0
```

Comment:
This comment is  confusing to me.  It sounds like it's saying we need to reset the offset back to 0 at the end of each block.

Reply:
This comment is referring to a previous implementation (actually is was a self-note). The `BlockCompressedInputStream.getFilePointer()` returns a virtual file pointer which includes the block-address and block-offset, and my first thought was that the block-offset is the file-offset; nevertheless, this is a pointer to the offset within the block, so for the end of the block the returned value for `BlockCompressedInputStream.getFilePointer()` is 0 for the block-offset (beginning of the next block). My first implementation was using `BlockCompressedFilePointerUtil `.

I removed the comment for something more informative now.